subprojects/core/src/main/java/org/gradle/caching/local/internal/DirectoryBuildCacheEntryExpiration.java

```diff
@@ -0,0 +1,61 @@
+/*
+ * Copyright 2024 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.caching.local.internal;
+
+import org.gradle.api.internal.cache.CacheConfigurationsInternal;
+import org.gradle.caching.BuildCacheServiceFactory;
+import org.gradle.caching.local.DirectoryBuildCache;
+import org.gradle.internal.time.TimeFormatting;
+import org.gradle.internal.time.TimestampSuppliers;
+
+import java.util.concurrent.TimeUnit;
+import java.util.function.Supplier;
+
+public class DirectoryBuildCacheEntryExpiration {
+    private final Supplier<Long> expiryTime;
+    private final String description;
+
+    public DirectoryBuildCacheEntryExpiration(DirectoryBuildCache directoryBuildCache, CacheConfigurationsInternal cacheConfigurations) {
+        @SuppressWarnings("deprecation")
+        int deprecatedRemoveUnusedEntriesAfter = directoryBuildCache.getRemoveUnusedEntriesAfterDays();
+
+        // Use the deprecated retention period if configured on `DirectoryBuildCache`, or use the core 'buildCache' cleanup configuration if not.
+        // If the deprecated property remains at the default, we can safely use the central value (which has the same default).
+        if (deprecatedRemoveUnusedEntriesAfter == CacheConfigurationsInternal.DEFAULT_MAX_AGE_IN_DAYS_FOR_BUILD_CACHE_ENTRIES) {
+            this.expiryTime = cacheConfigurations.getBuildCache().getRemoveUnusedEntriesOlderThanAsSupplier();
+            long expiryDuration = System.currentTimeMillis() - this.expiryTime.get();
+            long expiryDays = TimeUnit.MILLISECONDS.toDays(expiryDuration);
+            if (expiryDays >= 1) {
+                this.description = expiryDays + " days";
+            } else {
+                this.description = TimeFormatting.formatDurationTerse(expiryDuration);
+            }
+        } else {
+            this.expiryTime = TimestampSuppliers.daysAgo(deprecatedRemoveUnusedEntriesAfter);
+            this.description = deprecatedRemoveUnusedEntriesAfter + " days";
+        }
+
+    }
+
+    public Supplier<Long> getRemoveUnusedEntriesAfter() {
```

Comment:
Nitpick: I would use something less ambiguous, like `removeUnusedEntriesPriorTo()` or `removeUnusedEntriesOlderThan()` or even `retainUsedEntriesCreatedAfter()`.  I find `removeUnusedEntriedAfter()` to cause some mental gymnastics given it's in fact a timestamp.  It makes sense for a number of days, where we're always calculating back from the present, but when we're talking about a timestamp in the past, "after" seems to imply the wrong direction.